### PR TITLE
[Merged by Bors] - atx: verify syntactic correctness before fetching deps

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation
part of #4632

## Changes
break syntactic checks into 3 steps:
- do all the syntactically correctness checks on the current atx without its dependencies
- fetching atx dependencies, including poet proof, prev/positioning/commitment atxs
- do the syntactic checks based on the referenced atxs.

### CODE REVIEW

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
bors try

File: activation/handler.go [hunk 1]
```diff
@@ -165,47 +162,84 @@ func (h *Handler) ProcessAtx(ctx context.Context, atx *types.VerifiedActivationT
...
+		// Use the NIPost's Post metadata, while overriding the challenge to a zero challenge,
+		// as expected from the initial Post.
+		initialPostMetadata := *atx.NIPost.PostMetadata
+		initialPostMetadata.Challenge = shared.ZeroChallenge
+		if err := h.nipostValidator.Post(ctx, atx.SmesherID, *atx.CommitmentATX, atx.InitialPost, &initialPostMetadata, atx.NumUnits); err != nil {
+			return fmt.Errorf("invalid initial post: %w", err)
+		}
+		if err := h.nipostValidator.VRFNonce(atx.SmesherID, *atx.CommitmentATX, atx.VRFNonce, &initialPostMetadata, atx.NumUnits); err != nil {
+			return fmt.Errorf("invalid vrf nonce: %w", err)
+		}
```
Comment (poszu) [inline activation/handler.go]: NIT: The order of these checks could be switched as verifying VRF nonce should be much faster.

File: activation/handler.go [hunk 2]
```diff
@@ -165,47 +162,84 @@ func (h *Handler) ProcessAtx(ctx context.Context, atx *types.VerifiedActivationT
...
+	default:
+		if atx.InnerActivationTx.NodeID != nil {
+			return fmt.Errorf("prev atx declared, but node id is included")
+		}
+		if atx.InitialPost != nil {
+			return fmt.Errorf("prev atx declared, but initial post is included")
+		}
+		if atx.CommitmentATX != nil {
+			return fmt.Errorf("rpev atx declared, but commitment atx is included")
```
Comment (poszu) [inline activation/handler.go]:
```suggestion
			return fmt.Errorf("prev atx declared, but commitment atx is included")
```

File: activation/handler.go [hunk 3]
```diff
@@ -517,63 +517,54 @@ func (h *Handler) handleAtx(ctx context.Context, expHash types.Hash32, peer p2p.
...
 	existing, _ := h.cdb.GetAtxHeader(atx.ID())
 	if existing != nil {
-		logger.With().Debug("received known atx")
 		return fmt.Errorf("%w atx %s", errKnownAtx, atx.ID())
 	}
 
-	if atx.NIPost == nil {
-		return fmt.Errorf("nil nipst in gossip for atx %s", atx.ShortString())
+	if err := h.SyntacticallyValidate(ctx, &atx); err != nil {
+		return err
```
Comment (poszu) [inline activation/handler.go]: How about adding little context like for `...ValidateDeps()`?
```suggestion
		return fmt.Errorf("atx %v syntactically invalid: %w", atx.ShortString(), err)
```

File: activation/handler.go [hunk 4]
```diff
@@ -517,63 +517,54 @@ func (h *Handler) handleAtx(ctx context.Context, expHash types.Hash32, peer p2p.
...
-		)
+	if err := h.FetchReferences(ctx, &atx); err != nil {
+		return err
 	}
 
-	vAtx, err := h.SyntacticallyValidateAtx(ctx, &atx)
+	vAtx, err := h.SyntacticallyValidateDeps(ctx, &atx)
 	if err != nil {
-		return fmt.Errorf("received syntactically invalid atx %v: %w", atx.ShortString(), err)
+		return fmt.Errorf("atx %v syntatically invalid based on deps: %w", atx.ShortString(), err)
```
Comment (poszu) [inline activation/handler.go]:
```suggestion
		return fmt.Errorf("atx %v syntactically invalid based on deps: %w", atx.ShortString(), err)
```

File: activation/nipost_test.go [hunk 5]
```diff
@@ -412,16 +410,16 @@ func TestNIPostBuilder_BuildNIPost(t *testing.T) {
...
 	nipost, err = nb.BuildNIPost(context.Background(), &challenge2)
 	req.NoError(err)
 	req.NotNil(nipost)
 
 	// test state not loading if other challenge provided
 	poetDb.EXPECT().ValidateAndStore(gomock.Any(), gomock.Any()).Return(nil)
-	postProvider.EXPECT().GenerateProof(gomock.Any(), gomock.Any(), gomock.Any()).Times(1)
-	nipostValidator.EXPECT().Post(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Times(1).Return(nil)
+	postProvider.EXPECT().GenerateProof(gomock.Any(), gomock.Any(), gomock.Any())
+	nipostValidator.EXPECT().Post(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any())
```
Comment (poszu) [inline activation/nipost_test.go:424]: Why did you remove `.Times(1)`?

File: activation/validation.go [hunk 6]
```diff
@@ -63,8 +63,8 @@ func (v *Validator) NIPost(ctx context.Context, publishEpoch types.EpochID, node
 		return 0, err
 	}
 
-	if err := v.Post(ctx, publishEpoch, nodeId, commitmentAtxId, nipost.Post, nipost.PostMetadata, numUnits, opts...); err != nil {
-		return 0, fmt.Errorf("invalid Post: %w", err)
+	if err := v.Post(ctx, nodeId, commitmentAtxId, nipost.Post, nipost.PostMetadata, numUnits, opts...); err != nil {
+		return 0, fmt.Errorf("invalid Post: %v", err)
```
Comment (poszu) [inline activation/validation.go]: Why replace `%w` with `%v`? Shouldn't `%w` be used to signal the source of the error when bubbling up?

File: activation/validation.go [hunk 7]
```diff
@@ -170,17 +170,9 @@ func (*Validator) VRFNonce(nodeId types.NodeID, commitmentAtxId types.ATXID, vrf
...
-	}
-
-	if challenge.InitialPost == nil {
-		return fmt.Errorf("no prevATX declared, but initial Post is not included in challenge")
-	}
-
+func (v *Validator) InitialNIPostChallenge(challenge *types.NIPostChallenge, atxs atxProvider, goldenATXID types.ATXID) error {
 	if challenge.CommitmentATX == nil {
-		return fmt.Errorf("no prevATX declared, but commitmentATX is missing")
+		v.log.Panic("commitment atx is nil")
```
Comment (poszu) [inline activation/validation.go]: Why panic here? It opens a door to crash the node with a malicious ATX.

File: activation/validation.go [hunk 8]
```diff
@@ -218,21 +210,12 @@ func (*Validator) NIPostChallenge(challenge *types.NIPostChallenge, atxs atxProv
...
-	}
-
 	return nil
 }
 
-func (*Validator) PositioningAtx(id *types.ATXID, atxs atxProvider, goldenATXID types.ATXID, pubepoch types.EpochID, layersPerEpoch uint32) error {
+func (v *Validator) PositioningAtx(id *types.ATXID, atxs atxProvider, goldenATXID types.ATXID, pubepoch types.EpochID) error {
 	if *id == types.EmptyATXID {
-		return fmt.Errorf("empty positioning atx")
+		v.log.Panic("empty positioning atx")
```
Comment (poszu) [inline activation/validation.go]: ditto.

Btw, if `id` cannot be `nil`, then perhaps it should be passed by value to avoid the possibility of dereferencing a `nil` at all? The compiler should be able to optimize the copy away anyway.

File: activation/nipost_test.go (hunk 5 above)
Comment (countvonzero) [reply activation/nipost_test.go:424]: those are default. same as Return(nil).

File: activation/handler.go (hunk 3 above)
Comment (countvonzero) [reply activation/handler.go]: honestly i cannot think of a good context to add to the name. suggestions?

File: activation/validation.go (hunk 6 above)
Comment (countvonzero) [reply activation/validation.go]: oversight

File: activation/validation.go (hunk 8 above)
Comment (countvonzero) [reply activation/validation.go]: changed. thanks

File: activation/validation.go (hunk 7 above)
Comment (countvonzero) [reply activation/validation.go]: to check programming error. it's already verified that in SyntacticallyValidate() that this cannot be nil. but anyway. changed to return error

Comment (countvonzero) [reply activation/validation.go]: the idea was that all the nil were checked in the first stage of syntactic check. here it's doing checks on the deps. but not a big deal.

**Comment by countvonzero** [comment]
bors try

**Review by poszu — APPROVED** [review]

**Comment by countvonzero** [comment]
bors merge

**Comment by countvonzero** [comment]
bors merge
